openedx_learning/apps/authoring/publishing/models/draft_log.py

```diff
@@ -0,0 +1,237 @@
+       was created, but no PublishableEntityVersion was ever made for it, so
+       there was never a Draft version.
+    2. A Draft entry exists and points to a PublishableEntityVersion: This is
+       the most common state.
+    3. A Draft entry exists and points to a null version: This means a version
+       used to be the draft, but it's been functionally "deleted". The versions
+       still exist in our history, but we're done using it.
+
+    It would have saved a little space to add this data to the Published model
+    (and possibly call the combined model something else). Split Modulestore did
+    this with its active_versions table. I keep it separate here to get a better
+    separation of lifecycle events: i.e. this table *only* changes when drafts
+    are updated, not when publishing happens. The Published model only changes
+    when something is published.
+    """
+    # If we're removing a PublishableEntity entirely, also remove the Draft
+    # entry for it. This isn't a normal operation, but can happen if you're
+    # deleting an entire LearningPackage.
+    entity = models.OneToOneField(
+        PublishableEntity,
+        on_delete=models.CASCADE,
+        primary_key=True,
+    )
+    version = models.OneToOneField(
+        PublishableEntityVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+        blank=True,
+    )
+
+
+class DraftChangeLog(models.Model):
+    """
+    There is one row in this table for every time Drafts are created/modified.
+
+    There are some operations that affect many Drafts at once, such as
+    discarding changes (i.e. reset to the published versions) or doing an
+    import. These would be represented by one DraftChangeLog with many
+    DraftChangeLogRecords in it--one DraftChangeLogRecord for every
+    PublishableEntity that was modified.
+
+    Even if we're only directly changing the draft version of one
+    PublishableEntity, we will get multiple DraftChangeLogRecords if changing
+    that entity causes side-effects. See the docstrings for DraftChangeLogRecord
+    and DraftSideEffect for more details.
+    """
+    uuid = immutable_uuid_field()
+    learning_package = models.ForeignKey(LearningPackage, on_delete=models.CASCADE)
+    changed_at = manual_date_time_field()
+    changed_by = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+    )
+
+    class Meta:
+        verbose_name = _("Draft Change Log")
+        verbose_name_plural = _("Draft Change Logs")
+
+
+class DraftChangeLogRecord(models.Model):
+    """
+    A single change in the Draft version of a Publishable Entity.
+
+    We have one unusual convention here, which is that if we have a
+    DraftChangeLogRecord where the old_version == new_version, it means that a
+    Draft's defined version hasn't changed, but the data associated with the
+    Draft has changed because some other entity has changed.
```

Comment:
1. The current wording almost reads as if it's an accident or an edge case that old_version will sometimes equal new_version. I think that's unfair; it's really a solid model of what's happening and intuitively makes sense once you grok the whole system. Here's a suggested rewording:

```suggestion
    Changes often take form of a direct change to the content of the entity, which
    result in a bump of that entity's version, and thus new_version > old_version.
    However, this will not always be the case: if the data associated with the Draft
    has changed purely as a side effect of some other entity changing, then this will
    be represented here as a change log record where new_version == old_version.
```

2. Clarifying question: There will be instances where new_version > old_version, *and also* there's a side-effect record, right? As an example, any time an import includes parent-child relationships, I expect that a child and its parent can both have content changes. In other words: `new_version==old_version implies side-effect`, but not the other way around.